### PR TITLE
Phase 5: Fourier Feature Encoding — High-Frequency Input Enhancement (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Add learnable Fourier positional features to the input, enabling the model to resolve fine-scale pressure variations that standard coordinate features miss. The Transolver preprocesses 24-dim input features through an MLP — adding sin/cos features at multiple frequencies gives the MLP access to multi-scale spatial information from the start.

**Why this should work:**
- Tancik et al. (NeurIPS 2020) proved that standard MLPs have a spectral bias toward low-frequency functions. Fourier features break this bias by mapping coordinates to high-frequency sin/cos features.
- Surface pressure on airfoils has sharp spatial gradients (leading edge stagnation, suction peak, trailing edge recovery). These high-frequency patterns are exactly what Fourier features are designed to capture.
- The Transolver already has 4 fourier_freq parameters (logged to W&B) suggesting the codebase has some Fourier feature infrastructure — we should leverage and extend it.
- Zero architecture change to the Transolver itself. Only modifies the input preprocessing. Fully compatible with torch.compile.

**References:**
- Tancik et al., "Fourier Features Let Networks Learn High Frequency Functions in Low Dimensional Domains" (NeurIPS 2020)
- Mildenhall et al., "NeRF" (ECCV 2020) — positional encoding for 3D scene representation
- Li et al., "Fourier Neural Operator" — Fourier basis for PDE solving

## Instructions

See first comment for detailed implementation.

## Baseline (8-seed)
val/loss 0.404+/-0.004, p_in 13.33+/-0.58, p_oodc 8.37+/-0.22, p_tan 33.57+/-0.44, p_re 24.58+/-0.13

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results — Fourier Feature Encoding

### Implementation

Added configurable `FourierFeatureEncoding` module that maps spatial coordinates to sin/cos features at multiple frequencies. Replaces the existing inline Fourier PE (4+4 frequencies → 32 features) when `--fourier_features` is enabled.

New flags: `--fourier_features`, `--fourier_num_freqs`, `--fourier_learnable`, `--fourier_include_input`

### Results

| GPU | Config | val/loss | p_in | p_tan | p_oodc | p_re | Epoch |
|-----|--------|----------|------|-------|--------|------|-------|
| 0 | **4freq learnable** | **0.3976** | 13.5 | 33.4 | **7.9** | 24.4 | 159 |
| 1 | 8freq learnable | 0.4014 | 13.3 | 34.1 | 8.5 | **24.2** | 158 |
| 2 | 16freq learnable | 0.4043 | 13.7 | 34.0 | 8.6 | 24.4 | 154 |
| 3 | 8freq fixed | 0.4055 | 13.5 | 34.4 | 8.6 | 24.5 | 163 |
| 4 | 8freq no-include-input | 0.3997 | **12.9** | **33.3** | 8.6 | 24.3 | 160 |
| 5 | 32freq learnable | 0.4032 | 13.3 | **33.2** | 8.8 | 24.4 | 150 |
| 6 | Baseline seed 42 | 0.4013 | 12.8 | 33.8 | 8.3 | 24.5 | 159 |
| 7 | Baseline seed 43 | 0.3977 | 12.8 | 33.0 | 8.6 | 24.5 | 160 |
| | **Baseline mean** | **0.3995** | **12.8** | **33.4** | **8.45** | **24.5** | |

### Analysis

**Mixed results — no clear winner, but some interesting signals:**

1. **Best val/loss: 4freq learnable (0.3976)** — slightly better than baseline mean (0.3995), tied with baseline seed 43 (0.3977). Within noise.

2. **Best p_oodc: 4freq learnable (7.9)** — this is **-6.0% vs baseline mean 8.45**. This is the strongest OOD conditions result we've seen. The 4 learnable frequencies may capture the right spatial scales for extreme angle-of-attack/gap/stagger conditions.

3. **p_in slightly worse across all Fourier variants** — 12.9-13.7 vs 12.8 baseline. The Fourier features may be introducing noise for in-distribution predictions where the existing features are already sufficient.

4. **More frequencies ≠ better**: val/loss increases monotonically with num_freqs (0.3976 → 0.4014 → 0.4043 → 0.4032 for 4→8→16→32). The 4-freq variant matches the existing baseline's frequency count but uses a different initialization (log-linear 2^0...2^3 = 1,2,4,8 vs 0.5,2.0,8.0,32.0 fixed + 1.0,3.0,6.0,16.0 learned).

5. **Fixed vs learnable**: Fixed frequencies (GPU3, 0.4055) slightly worse than learnable (GPU1, 0.4014), confirming the model benefits from adapting the frequency spectrum.

6. **No include_input (GPU4, 0.3997)**: Dropping the raw coords from the Fourier output gives the 2nd best val/loss and best p_in (12.9). The raw xy coords may be redundant when Fourier features are present.

**The p_oodc improvement (7.9) on the 4freq variant is the most interesting finding** — it suggests that different Fourier frequency initialization may help with OOD generalization. However, this is a single-seed result and would need multi-seed validation.

### Commands Used

```bash
CUDA_VISIBLE_DEVICES=<N> python train.py --agent frieren --wandb_group "phase5/fourier-features" \
  --fourier_features --fourier_num_freqs <4|8|16|32> [--fourier_learnable False] [--fourier_include_input False] \
  --field_decoder --adaln_output --use_lion --lr 2e-4 --cosine_T_max 180 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --disable_pcgrad --pressure_first --pressure_deep --seed 42
```

### Suggested Follow-ups

1. **Multi-seed validation of 4freq learnable** — the p_oodc=7.9 result needs validation. Run seeds 42-47 for the 4-freq config.
2. **Try 4freq no-include-input** — combine the two best ideas (4 freqs + no raw coords).
3. **Different frequency initialization** — the current log-linear spacing (1,2,4,8) differs from the baseline's hand-tuned (0.5,2,8,32). The initialization matters more than the count.

**Comment by tcapelle** [comment]
ADVISOR: ## Review — Fourier Feature Encoding

**Result: CLOSE — Null for p_in (primary metric), interesting p_oodc lead but p_in trades off.**

Key findings:
1. **More frequencies = worse**: Consistent with the model having limited epochs to learn high-frequency features. 4-freq is optimal.
2. **p_oodc=7.9 from 4-freq** is below the 8-seed baseline range (7.99-8.74) — genuinely interesting for OOD generalization
3. **But p_in=13.5 vs 12.8 baseline** — a clear regression on our primary metric. Can't merge an improvement that hurts the main metric.
4. The frequency initialization insight is useful: log-linear 1,2,4,8 differs from baseline's 0.5,2,8,32

The existing Fourier PE in the codebase already provides similar functionality. The learnable aspect is interesting but doesn't overcome the p_in regression. Closing.

Good systematic sweep, frieren. The frequency analysis and the no-include-input finding are useful for understanding what information the model uses.
